dvdisaster/documentation/dvdisaster.de.1
.TH DVDISASTER 1 "2010-02-07" "0.80" "Schutz f\[:u]r CD/DVD Medien"

.SH NAME
DVDISASTER \- Schutz f\[:u]r CD/DVD Medien gegen Datenverlust durch Alterung
oder Kratzer

.SH SYNTAX
.B dvdisaster
.RB [\| \-r \||\| \-c \||\| \-f \||\| \-s \||\| \-t[q] \||\| \-u \|]
.RB [\| \-d
.IR device \|]
.RB [\| \-p
.IR prefix \|]
.RB [\| \-i
.IR image \|]
.RB [\| \-e
.IR eccfile \|]
.RB [\| \-o
.IR file|image \|]
.RB [\| \-a
.IR codec-list \|]
.RB [\| \-j
.IR n \|]
.RB [\| -n
.IR n% \|]
.RB [\| -m
.IR n \|]
.RB [\| -v \|]
.RB [\| -x
.IR n \|]
.RB [\| \-\-adaptive-read \|]
.RB [\| \-\-auto-suffix \|]
.RB [\| \-\-cache-size
.IR n \|]
.RB [\| \-\-dao \|]
.RB [\| \-\-defective-dump \|
.IR d \|]
.RB [\| \-\-driver \|
.IR d \|]
.RB [\| \-\-eject \|]
.RB [\| \-\-fill-unreadable
.IR n \|]
.RB [\| \-\-ignore-fatal-sense \|]
.RB [\| \-\-internal-rereads
.IR n \|]
.RB [\| \-\-prefetch-sectors
.IR n \|]
.RB [\| \-\-raw-mode
.IR n \|]
.RB [\| \-\-read-attempts
.IR n-m \|]
.RB [\| \-\-read-medium
.IR n \|]
.RB [\| \-\-read-raw \|]
.RB [\| \-\-speed-warning
.IR n \|]
.RB [\| \-\-spinup\-delay
.IR n \|]

.SH BESCHREIBUNG
.B DVDISASTER
erzeugt einen Sicherheitspuffer gegen Datenverluste, die auf CD- und
DVD-Datentr\[:a]gern durch Alterung oder Kratzer entstehen. Es erzeugt
Fehlerkorrekturdaten, um bei nachfolgenden Datentr\[:a]ger-Problemen unlesbare
Sektoren zu rekonstruieren.

.SH TYPISCHE ANWENDUNGEN

.TP
\fBdvdisaster\fP \fB-d\fP \fI/dev/hdc\fP \fB-s\fP
\[:U]berpr\[:u]ft den Datentr\[:a]ger in Laufwerk \fI/dev/hdc\fP auf Fehler.
.TP
\fBdvdisaster\fP \fB-d\fP \fI/dev/hdc\fP \fB-i\fP \fIabbild.iso\fP \fB-r\fP
Erstellt ein Abbild vom Datentr\[:a]ger in Laufwerk \fI/dev/hdc\fP als Datei \fIabbild.iso\fP.
.TP
\fBdvdisaster\fP \fB-d\fP \fI/dev/hdc\fP \fB-i\fP \fImedium.iso\fP \fB--read-raw\fP \fB-r\fP
Erstellt ein Abbild wie oben beschrieben. Die Sektoren werden anhand ihrer
EDC- und L-EC-Rohdaten auf Unversehrtheit \[:u]berpr\[:u]ft. Dies ist nur f\[:u]r 
CD-Datentr\[:a]ger m\[:o]glich; anderenfalls hat die \fB--read-raw\fP-Option
keine Wirkung.
.TP
\fBdvdisaster\fP \fB-d\fP \fI/dev/hdc\fP \fB-i\fP \fImedium.iso\fP \fB--read-attempts\fP \fIn-m\fP \fB-r\fP
Erstellt ein Abbild wie oben beschrieben. F\[:u]r besch\[:a]digte Sektoren werden
mindestens \fIn\fP und h\[:o]chstens \fIm\fP Leseversuche unternommen. Die
Lesbarkeit von besch\[:a]digten CD-Datentr\[:a]gern kann in Verbindung mit
\fB--read-raw\fI besser werden.
.TP
\fBdvdisaster\fP \fB-d\fP \fI/dev/hdc\fP \fB-i\fP \fIabbild.iso\fP \fB-e\fP \fIkorr.ecc\fP \fB-r\fP \fB--adaptive-read\fP
Verwendet die angepasste Lesestrategie, um ein Abbild vom 
Datentr\[:a]ger in Laufwerk \fI/dev/hdc\fP als Datei \fIabbild.iso\fP
zu erstellen. Der Lesevorgang wird automatisch beendet
sobald gen\[:u]gend Daten eingelesen wurden, um das Abbild mit Hilfe der 
Fehlerkorrektur-Datei \fIkorr.ecc\fP wiederherzustellen:
.TP
\fBdvdisaster\fP \fB-i\fP \fIabbild.iso\fP \fB-e\fP \fIkorr.ecc\fP \fB-c\fP
Erzeugt eine Fehlerkorrektur-Datei \fIkorr.ecc\fP f\[:u]r das Abbild \fIabbild.iso\fP.
.TP
\fBdvdisaster\fp \fB-i\fP \fIabbild.iso\fP \fB-m\fIRS02\fP \fB-n\fP \fI350000\fP \fB-c\fP
Erweitert ein Abbild \fIabbild.iso\fP mit Fehlerkorrektur-Daten,
wobei das Abbild auf nicht mehr als \fI350000\fP Sektoren vergr\[:o]ssert wird.
Ohne Angabe von \fB-n\fP wird das Abbild auf die Gr\[:o]sse des 
kleinstm\[:o]glichen Datentr\[:a]gers (CD, DVD, DVD9) erweitert.
Beachten Sie das fehlende Leerzeichen zwischen \fB-m\fP und \fIRS02\fP.
.TP
\fBdvdisaster\fP \fB-i\fP \fIabbild.iso\fP \fB-e\fP \fIkorr.ecc\fP \fB-f\fP
Wiederherstellung der Abbild-Datei \fIabbild.iso\fP 
mit Hilfe der Fehlerkorrektur-Datei \fIkorr.ecc\fP.
.TP
\fBdvdisaster\fP \fB-i\fP \fIabbild.iso\fP \fB-e\fP \fIkorr.ecc\fP \fB-t\fP
\[:U]berpr\[:u]ft den Inhalt des Abbilds \fIabbild.iso\fP 
mit Hilfe der Fehlerkorrektur-Datei \fIkorr.ecc\fP auf Vollst\[:a]ndigkeit und Korrektheit.
.TP
\fBHINWEIS:\fP
Geben Sie in den obigen Beispielen  die \fB-e korr.ecc\fP-Optionen nicht an, 
wenn Sie mit erweiterten Abbildern arbeiten.

.SH OPTIONEN
Aktions-Auswahl (mindestens eine Aktion muss gew\[:a]hlt werden):
.TP
.B \-r, \-\-read
Datentr\[:a]ger-Abbild auf Festplatte einlesen. -rn-m liest einen Bereich von
Sektoren, z.B. -r100-200.
.TP
.B \-c, \-\-create
Erzeugt Fehlerkorrektur-Daten f\[:u]r das Abbild.
.TP
.B \-f, \-\-fix
Versucht das Abbild mit Hilfe der Fehlerkorrektur-Daten zu reparieren.
.TP
.B \-s, \-\-scan
Untersucht den Datentr\[:a]ger auf Lesefehler.
.TP
.B \-t, \-\-test, \-tq, \-test=q
Pr\[:u]ft die Struktur der .iso und .ecc - Dateien.
Bei Angabe der "q"-Option werden nur die Informationen ausgegeben,
die ohne vollständiges Lesen der Dateien ermittelt werden k\[:o]nnen.
.TP
.B \-u, \-\-unlink
L\[:o]scht .iso - Dateien nach Beenden der vorherigen Aktionen.
.PP

Laufwerks- und Dateinamen:
.TP
.B \-d, \-\-device device
Von diesem Laufwerk lesen (Standard: /dev/cdrom).
.TP
.B \-p, \-\-prefix prefix
Anfang der .iso/.ecc - Dateien (Standard: medium.* ).
.TP
.B \-i, \-\-image imagefile
Name der Abbild - Datei (Standard: medium.iso).
.TP
.B \-e, \-\-ecc eccfile
Name der Fehlerkorrektur-Datei (Standard: medium.ecc).
.TP
.B \-o, \-\-ecc-target file|image
Gibt an ob mit RS03 Fehlerkorrektur-Dateien oder erweiterte
Abbilder erzeugt werden sollen (Standard: erweiterte Abbilder).
.PP

Feineinstellung (Beachten Sie die Hinweise in der Dokumentation!):
.TP
.B \-a, \-\-assume codec1,codec2,...
Nimmt an daß das Abbild mit einem der angegebenen Kodierer erweitert
wurde. Dies bewirkt eine erschöpfende Suche nach den Datenstrukturen
der Kodierer und ist hilfreich, um in beschädigten Datenträgern enthaltene
Fehlerkorrektur-Daten wiederzufinden. Wenn das Abbild allerdings keine
der angegebenen Fehlerkorrektur-Daten enthält, verschwendet diese Option
eine große Menge Rechenzeit und Ein-/Ausgabeoperationen.
.RS
M\[:o]gliche Werte sind RS02 und RS03.
.RE
.TP
.B \-j, \-\-jump n
\[:u]berspringe n Sektoren nach einem Lesefehler (Standard: 16).
.TP
.B \-n, \-\-redundancy n%
Redundanz der Fehlerkorrektur (Standard: 14.3%).
.TP
.B \-m, \-\-method n
Fehlerkorrektur-Methoden anzeigen/ausw\[:a]hlen (Standard: RS01).
.RS
M\[:o]gliche Werte sind RS01 und RS02.
.RE
.TP
.B \-v, \-\-verbose n%
mehr erl\[:a]uternde Ausgaben
.TP
.B \-x, \-\-threads n
Verwende n Kontrollfäden für den RS03-Kodierer/Dekodierer. Empfohlen
sind 2 bzw. 4 Kontrollfäden für 2- bzw. 4-Kern-Prozessoren. Lassen Sie
auf größeren Systemen einen Kontrollfaden für Verwaltungszwecke frei,
d.h. benutzen Sie 7 Kontrollfäden auf einem 8-Kern-System.
.TP
.B \-\-adaptive-read
verwende optimierte Lesestrategie f\[:u]r defekte Datentr\[:a]ger.
.TP
.B \-\-auto-suffix
automatisches Anf\[:u]gen der .iso- und .ecc-Dateiendungen.
.TP
.B \-\-cache-size n
Zwischenspeicher in MB bei .ecc-Datei-Erzeugung - (Standard: 32MB).
.TP
.B \-\-dao
unterstelle DAO; Abbild am Ende nicht k\[:u]rzen.
.TP
.B \-\-defective-dump d
Gibt das Unterverzeichnis zum Sammeln von unvollständigen
Roh-Sektoren an.
.TP
.B \-\-driver d  (nur für Linux)
dvdisaster verwendet unter Linux normalerweise den Treiber für optische
Laufwerke. Einige sehr alte SCSI-Kontroller erzeugen dabei Systemabstürze. 
Probieren Sie in diesem Fall mit
\-\-driver=sg das Laufwerk über den alternativen SCSI-Treiber anzusteuern.
.TP
.B \-\-eject
Datentr\[:a]ger nach erfolgreichem Lesen auswerfen.
.TP
.B \-\-fill-unreadable n
f\[:u]lle unlesbare Sektoren mit Byte n.
.TP
.B \-\-ignore-fatal-sense
Lesen nach m\[:o]glicherweise schwerwiegenden Fehlern fortsetzen.
.TP
.B \-\-internal-rereads n
Leseversuche innerhalb des Laufwerks f\[:u]r besch\[:a]digte CD-Sektoren (Standard: -1)
.RS
Laufwerke unternehmen normalerweise mehrere Versuche um einen besch\[:a]digten Sektor 
zu lesen. Es ist typischerweise g\[:u]nstiger diesen Wert auf 0 oder 1 zu setzen
und die Anzahl der Leseversuche \[:u]ber den Parameter --read-attempts zu steuern.
Viele Laufwerke ignorieren diese Einstellung ohnehin. Benutzen Sie den Wert -1 um 
die Standardeinstellungen des Laufwerks zu verwenden.
.RE
.TP
.B \-\-prefetch-sectors n
n Sektoren für die RS03-(De)kodierung im Voraus laden (Standard: 32)
.RS
Ein Wert von n verbraucht ungefähr n MB Arbeitsspeicher.
.RE
.TP
.B \-\-raw-mode n
"Raw"-Lese-Verfahren f\[:u]r besch\[:a]digte CD-Sektoren (default: 20)
.RS
Das empfohlene Verfahren ist 20, bei dem das Laufwerk die eingebaute Fehlerkorrektur
so weit wie m\[:o]glich anwendet, bevor es einen besch\[:a]digten Sektor 
zur\[:u]ckgibt. Einige Laufwerke k\[:o]nnen besch\[:a]digte Sektoren allerdings
nur in der Betriebsart 21 lesen. Dabei wird die letzte Stufe der eingebauten
Fehlerkorrektur nicht ausgef\[:u]hrt und der Sektor wird unkorrigiert
zur\[:u]ckgegeben.
.RE
.TP
.B \-\-read-attempts n-m
versucht einen besch\[:a]digten Sektor n bis m-mal zu lesen.
.TP
.B \-\-read-medium n
liest den gesamten Datentr\[:a]ger bis zu n-mal.
.TP
.B \-\-read-raw
liest in der "raw"-Betriebsart sofern m\[:o]glich.
.TP
.B \-\-speed-warning n

dvdisaster/CHANGELOG

0.79.1 07-02-2010	
       A word of caution: This version is still evolving and some parts
       are not yet implemented. It may contain severe bugs and fail in 
       non-obvious ways, even in functions which worked in previous versions. 
       Do not process important data with this version and do
       not keep images and error correction data for archival purposes;
       that's what the stable version 0.72 is for.

	Semantic Changes
	- query-size option replaced with -a/--assume option.
	  command line default is now NOT to do exhaustive search
	  for RS02 information; add -a RS02 to enable it
	  (old behaviour: exhaustive check was on; required
	   --query-size=udf to disable it)
	- same changes in the preferences dialogues: 
	  "Image size" field replaced with 
	  "Error correction data recognization" options.

	Changes
	- added MIN_TRANSFER_LEN and length sanity check as workaround 
	  for buggy drive chipsets (thanks to R.G. for reporting this
	  and helping with debugging output)
	- added -tq command line option as suggested by Mikhail Arefiev
	- fixed alignment brokeness between 32bit and 64bit systems
	  in ecc header and crc sectors
	- added RS03 encoding (GUI)
	- added RS03 verifying (CLI and GUI)
	- fixed RS02/RS03 crash when encoding invoked without image file
	- added commandline RS03 encoding (singlethreaded/multithreaded)
	- added RS03 decoding and verifying (CLI and GUI)
	- fixed "typical uses" links in welcome screen (Thanks to C. Steigies)
	- added warning when same files are selected for .iso
	  and error correction data (Thanks to A. Postkutscher
	  for suggesting this!)
	- removed ASPI driver and ASPI/SPTI splitting logic
	- removed 2GB file splitting

0.72 (pl-1) 08-08-2009
	- Pablo Almeida provided a Portuguese translation
	  for the screen text.
	- workaround for freeze on Win XP for certain drive/CD-RW pairs

0.72 04-07-2009
	- Igor Gorbounov completed the Russian translation
	- fixed lockups in the spiral under Mac OS X
	- fixed crash when calling dvdisaster on command line without
	  specifying a file name on Mac OS X

0.72 (rc-2)
	- fixed lockups in the spiral under Mac OS X
	- fixed crash when calling dvdisaster on command line without
	  specifying a file name on Mac OS X

0.72 (rc-1) 11-04-2009
        - Documentation and locale cleanup for en, de
	- added app bundle creation for Mac OS X
	- Graphical bug: red/green sector markings were incorrect when doing
	  linear reading with high sector skip numbers. Fixed.
	  Speed curve will no longer travel below the x axis.
	- fixed rare crash while linear reading due to a race condition
	  by improper ClearCrcCache() usage.
	- fixed bogus dead sector warning in the linear reader 
	  after encountering unreadable sectors
	- made adaptive reading pursue raw reading more aggressively;
	  effectively trading in place re-reads for seeks over
	  the medium.
	- --help output for --defective-dump was complete nonsense. Fixed.
	- detection of read errors at the end of track at once disks
	  would sometimes cut off too many sectors when doing multiple
	  reading passes and a carefully crafted number of extra read
	  errors occurred. Fixed.
	- fixed crash in command line mode when doing multiple 
	  reading passes.

------ 0.72 released ------

0.71 (devel-28) 08-03-2009
	- changed SSE2 handling; now only the interesting part of
	  dvdisaster are compiled with -msse2 and selection between
	  normal and SSE2 code versions are done at run time.
	  (Don't waste your time searching for SSE2 code in 0.7x -
	   there is none; but we need the option to configure for it
	   in preparation of 0.80)
	   Thanks to Karsten H., Nathan S. and Tobias G. for bringing 
	   this issue to my attention!
	- changed treatment of CRC failures: old behaviour was to 
	  store a defective sector marker; now the actual sector
	  contents are written to the image file.
	- starting a scan/after automatically loading a medium
	  would not always work. Fixed.
	- command line mode will no longer scan for available 
	  devices when one is already given via -d.
	  Thanks to Raphael M. for noting this.
	- fixed build failure on FreeBSD; updated atapicam loading
	  instructions.
	- opening the preferences dialog produced a GTK warning. Fixed.
        - log window would crash during update or when closed and
	  re-opened repeatedly. Fixed the bunch of race conditions
	  causing this.

dvdisaster/documentation/include/dvdisaster.php
<?php

# dvdisaster: Homepage layout functions
# Copyright (C) 2007-2009 Carsten Gnörlich

require("version.php");

# Preset some global variables

$project_at_hoster="http://sourceforge.net/projects/dvdisaster";
$max_news_flash_items = 7;
$create_feed = 0;

# Find out from where we have been called;
# the file name is important for creation of the index.
# Also we extract the current locale from the path which is supposed
# to be .../LC_php/file.php. Maybe this is too hackish...

$script_path = current(get_included_files());
$script_file = basename($script_path);
$script_name = basename($script_path, ".php");
$script_dir  = dirname($script_path);
$script_lang = substr($script_dir, strlen($script_dir)-2, 2);

# Needed to exatract some meaningful title text from the toc.php

$toc_title_mode = 0;
$toc_title_content = "RBG";

# Load the appropriate localization file

require("dict_" . $script_lang . ".php");

# Locale wrappers for toc.php stubs

function de($msg) 
{  global $toc_title_mode; 

   if($toc_title_mode == 1) toc_title($msg, "de"); 
   else                     toc_link($msg, "de"); 
};

function en($msg) 
{  global $toc_title_mode; 

   if($toc_title_mode == 1) toc_title($msg, "en"); 
   else                     toc_link($msg, "en");
};

function ru($msg) 
{  global $toc_title_mode; 

   if($toc_title_mode == 1) toc_title($msg, "ru"); 
   else                     toc_link($msg, "ru");
};

#
# Create the HTML header ----------------------------------------------------------
#

function start_html()
{  global $toc_title_mode;
   global $toc_title_content;
   global $script_name;
   global $script_lang;
   global $create_feed;
   global $trans_atom_title;

   echo "<html>\n";
   echo "<head>\n";
   echo " <meta http-equiv=\"content-type\" content=\"text/html; charset=UTF-8\">\n";
   $toc_title_mode = 1;
   $toc_title_content = "dvdisaster";
   require("toc.php");
   echo " <title>$toc_title_content</title>\n";
   $toc_title_mode = 0;
   echo " <link rel=\"stylesheet\" type=\"text/css\" href=\"../include/dvdisaster.css\">\n";
   if(!strcmp($script_name, "index"))
   {  echo "<link rel=\"alternate\" type=\"application/atom+xml\" href=\"http://dvdisaster.net/$script_lang/feed/atom.xml\" title=\"$trans_atom_title\" />\n";
      $create_feed=1;
   }

   echo "</head>\n";
   echo "<body>\n";
}

###
### Head bar ------------------------------------------------------------------------
###
# This includes the "dvdisaster version ..." header. Depending on the mode we
# - provide the link to the online version for local documentation, or
# - create the language switch for the online publication.

function lang_link($lang_name, $lang, $spacing)
{  global $script_file;
   global $script_lang;
   
   if($spacing) $space="&nbsp;&nbsp;&nbsp;";
   else         $space="";

   if(strcmp($lang, $script_lang)) 
        echo "       <a href=\"../${lang}/$script_file\">$lang_name</a> $space\n";
   else echo "       $lang_name $space\n";
}

function begin_page()
{  global $cooked_version;
   global $have_experimental;
   global $stable_version;
   global $trans_to_hoster;
   global $trans_to_internet;
   global $trans_version;
   global $project_at_hoster;
   global $script_lang;
   global $mode;

   # Begin HTML file

   start_html();

   # Write the title header

   echo "\n<!--- Title header --->\n";
   echo "<table width=\"100%\" cellpadding=\"0\" border=\"0\">\n";
   echo "  <tr>\n";
   echo "     <td align=\"left\">\n";
   echo "       <font size=\"+3\"><b>dvdisaster</b></font>\n";
   if(!strcmp($have_experimental, "no"))
      echo "       <i>$trans_version $cooked_version</i>\n";
   else
      echo "       <i>$trans_version $stable_version / $cooked_version</i>\n";
   if(!strcmp($mode, "local"))
   {  echo "  </td>\n";
      echo "  <td align=\"right\">\n";
      echo "     <font size=\"+3\">&nbsp;</font><a href=\"http://dvdisaster.net/$script_lang/\">$trans_to_internet</a>\n";
      lang_link("", "de", 0); # TODO: This is a quick hack
      lang_link("", "en", 0); # to produce all locales for
      lang_link("", "ru", 0); # Windows. Do it better!
   }
   echo "     </td>\n";
   echo "  </tr>\n";
   echo "</table>\n";

   # Write the language chooser depending on mode

   echo "\n<!--- Language chooser or separator --->\n";
   echo "<table width=\"100%\" cellpadding=\"0\" border=\"0\">\n";

   echo "  <tr bgcolor=\"#000000\">\n";
   echo "    <td colspan=\"2\" width=\"100%\"><img width=1 height=1 alt=\"\"></td>\n";
   echo "  </tr>\n";

   if(strcmp($mode, "local"))
   {  
      echo "  <tr>\n";
      echo "    <td align=\"left\"><a href=\"$project_at_hoster\">$trans_to_hoster</a></td>\n";
      echo "    <td align=\"right\">\n";
#      lang_link("&#268;esky", "cs");
      lang_link("Deutsch", "de", 1);
      lang_link("English", "en", 1);
      lang_link("Russian", "ru", 0);
      echo "    </td>\n";
      echo "  </tr>\n";
      echo "  <tr bgcolor=\"#000000\">\n";
      echo "    <td colspan=\"2\" width=\"100%\"><img width=1 height=1 alt=\"\"></td>\n";
      echo "  </tr>\n";
   }

   # Both modes get the separator 

   echo "  <tr><td colspan=\"2\" width=\"100%\" height=\"10\">\n";
   echo "     <img width=1 height=1 alt=\"\">\n";
   echo "  </td></tr>\n";
   echo "</table>\n";

   # Insert the navigation column

   create_navigation();
}

###
### Navigation column ----------------------------------------------------------------
###

#
# Helper functions for creating the table of contents
#

function section($section_name)
{  global $toc_section;
   global $toc_mode;

   $toc_section = $section_name;
   $toc_mode = "section";
}

function subsection($subsection_name)
{  global $toc_subsection;
   global $toc_mode;

   $toc_subsection = $subsection_name;
   $toc_mode = "subsection";
}

function subsubsection($subsubsection_name)
{  global $toc_subsubsection;
   global $toc_mode;

   $toc_subsubsection = $subsubsection_name;
   $toc_mode = "subsubsection";
}

function toc_title($msg, $lang)
{  global $script_lang;
   global $script_name;
   global $toc_mode;
   global $toc_section;
   global $toc_subsection;
   global $toc_subsubsection;
   global $toc_title_content;

   if(strcmp($lang, $script_lang)) return; # wrong locale

   if(!strcmp($toc_mode, "section") && !strcmp($toc_section, $script_name))
     $toc_title_content = $msg;

   if(!strcmp($toc_mode, "subsection") && !strcmp($toc_subsection, $script_name))
     $toc_title_content = $msg;

   if(!strcmp($toc_mode, "subsubsection") && !strcmp($toc_subsubsection, $script_name))
     $toc_title_content = $msg;
}

function toc_link($msg, $lang)
{  static $separator=0;
   global $script_lang;
   global $script_name;
   global $toc_section;
   global $toc_subsection;
   global $toc_subsubsection;
   global $toc_mode;

   if(strcmp($lang, $script_lang)) return; # wrong locale

   # Decide whether this is the currently unfolded section
   # and render it accordingly

   if(!strcmp($toc_mode, "section"))
   {  $target=$toc_section . ".php";

      # Draw the separator between sections (except for the first one)

      if(!$separator) $separator=1;
      else echo "            <tr><td></td><td></td><td height=\"10\"></td></tr>\n";

      if(!strcmp($toc_section, $script_name))
	   echo "            <tr><td colspan=3><font size=\"-1\">$msg</font></td></tr>\n";
      else echo "            <tr><td colspan=3><font size=\"-1\"><a href=\"$target\">$msg</a></font></td></tr>\n";
   }

   if(   !strcmp($toc_mode, "subsection") 
      && !strncmp($toc_section, $script_name, strlen($toc_section)))
   {  $target=$toc_subsection . ".php";

      if(!strcmp($toc_subsection, $script_name))
	   echo "            <tr><td valign=\"top\" width=\"1%\">&middot;</td><td colspan=2><font size=\"-1\">$msg</font></td></tr>\n";
      else echo "            <tr><td valign=\"top\" width=\"1%\">&middot;</td><td colspan=2><font size=\"-1\"><a href=\"$target\">$msg</a></font></td></tr>\n";
   }

   # Using strlen($toc_section)+1 is inconvenient as it hardcodes the
   # section/subsection10/subsubsection11 scheme. Improve!

   if(   !strcmp($toc_mode, "subsubsection") 
      && !strncmp($toc_subsection, $script_name, strlen($toc_section)+1))
   {  $target=$toc_subsubsection . ".php";

      if(!strcmp($toc_subsubsection, $script_name))
	   echo "            <tr><td valign=\"top\"></td><td>-</td><td><font size=\"-1\">${msg}</font></td></tr>\n";
      else echo "        <tr><td valign=\"top\"></td><td>-</td><td><font size=\"-1\"><a href=\"$target\">$msg</a></font></td></tr>\n";
   }

}

#
# Helper functions for creating the news pages and -flash
#

function news_headline($headline)
{  global $news_flash;
   global $atom_handle;
   global $create_feed;
   global $script_lang;
   global $doc_dir;
   global $trans_atom_title;

   if(!$news_flash) echo "    <h3>$headline</h3>\n";

   if($create_feed != 1) return;

   # Prodoce atom feed xml file

   $atom_name="atom.xml";
   $atom_handle=fopen("$doc_dir/$script_lang/feed/$atom_name","w");
   
   fwrite($atom_handle, "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n");
   fwrite($atom_handle, "<feed xmlns=\"http://www.w3.org/2005/Atom\">\n");
   fwrite($atom_handle, "<id>tag:dvdisaster.net,2009-10-02:/$script_lang/feeds/$atom_name</id>\n");
   fwrite($atom_handle, "<title>$trans_atom_title</title>\n");
   $updated=date(DATE_ATOM);
   fwrite($atom_handle, "<updated>$updated</updated>\n");
   fwrite($atom_handle, "<link rel=\"self\" href=\"http://dvdisaster.net/$script_lang/feed/$atom_name\" type=\"application/atom+xml\" />\n");
   fwrite($atom_handle, "<author>\n");
   fwrite($atom_handle, " <name>Carsten Gnörlich</name>\n");
   fwrite($atom_handle, " <uri>http://www.dvdisaster.org</uri>\n");
   fwrite($atom_handle, "</author>\n");
}

function news_finalize()
{  global $atom_handle;
   global $create_feed;

   if($create_feed != 1)
     return;

   fwrite($atom_handle, "</feed>\n");
   fclose($atom_handle);
}

function news_item($date, $headline, $body, $atom_tag, $atom_created, $atom_updated)
{  global $news_flash;
   global $news_counter;
   global $max_news_flash_items;
   global $atom_handle;
   global $create_feed;
   global $script_lang;

   $news_counter++;

   if($create_feed == 1)
   {  $stripped=strtr(strip_tags($body),"\n"," ");
      $summary=substr($stripped, 0, 240);
      $cutpos=240-strlen(strrchr($summary, " "));
      $summary=substr($stripped, 0, $cutpos)." [...]";      

      fwrite($atom_handle,"<entry>\n");
      fwrite($atom_handle,"<title>$headline</title>\n");
      fwrite($atom_handle,"<category term=\"News\"/>\n"); 
      $created=substr($atom_created,0,10);
      fwrite($atom_handle,"<id>tag:dvdisaster.net,$created:/$script_lang/news.html/$atom_tag</id>\n");
      fwrite($atom_handle,"<published>$atom_created</published>\n");
      fwrite($atom_handle,"<updated>$atom_updated</updated>\n");
      fwrite($atom_handle,"<link href=\"http://dvdisaster.net/$script_lang/news.html#item$atom_tag\"/>\n");
      fwrite($atom_handle,"<summary>$summary</summary>\n");
      fwrite($atom_handle,"</entry>\n");
   }

   if($news_flash)
   {  if($news_counter > $max_news_flash_items)
        return;

      echo "          <font size=\"-1\">$date</font> <br>\n";
      echo "          <font size=\"-1\">\n";
      echo "            <a href=\"news.php#item$atom_tag\">$headline</a>\n";
      echo "          </font><p>\n";
   }
   else
   {  
      echo "    <table width=\"90%\">\n";
      echo "      <tr>\n";
      echo "        <td><a name=\"item$atom_tag\"></a><b>${headline}</b></td>\n";
      echo "        <td align=\"right\">$date</td>\n";
      echo "      </tr>\n";
      echo "    </table>\n";
?>
    <table width="90%" cellpadding="0" cellspacing="0">
       <tr bgcolor="#000000" height=1>
	 <td width="100%" height=1><img width=1 height=1 alt=""></td>
       </tr>
    </table>
    <table width="90%" cellpadding="0" cellspacing="0">
       <tr bgcolor="#ffffff" height=5>
         <td width="100%" height=1><img width=1 height=1 alt=""></td>
       </tr>
    </table>
<?php
      echo "    <table width=\"90%\">\n";
      echo "      <tr><td>\n";
      echo "$body\n";
      echo "      </td></tr>\n";
      echo "    </table>\n";
      echo "    <pre> </pre>\n";
   }
}

#
# Actual navigation assembly
#

function create_navigation()
{  global $trans_contents;
   global $script_file;
   global $mode;   

   if(!strcmp($script_file, "index.php"))
        $body_width = "57%";
   else $body_width = "77%";
?>
  
<!--- Main body (Navigation, actual page content, optional news column) ---> 
<table width="100%" cellspacing=0>
 <tr>
  <!--- Navigation --->
  <td bgcolor="#f0f0f0" valign="top" width="20%">
    <table width="100%" cellpadding="10">
      <tr>
        <td>
<?php
echo "         <font size=\"-1\"><b>$trans_contents</b></font>\n";
?>
          <table width="100%" cellpadding="0" cellspacing="0">
	    <tr bgcolor="#000000">
	      <td width="100%"><img width=1 height=1 alt=""></td>
	    </tr>
	  </table><p>

          <table width="100%">
<?php require("toc.php");?>
	  </table>
        </td>
      </tr>
    </table>
  </td>
  <td></td>

  <!--- Actual page contents --->
<?php
echo "  <td valign=\"top\" width=\"$body_width\" rowspan=\"2\">\n";

  # body contents must be appended here from calling page
}

###
### Footer ----------------------------------------------------------------------------
### 
# Closes the body table. Optionally includes the news flash.
# Appends the copyright and redistribution terms.
#

function end_page()
{  global $script_file;
   global $trans_copyright;
   global $trans_modified;
   global $trans_news;
   global $trans_fdl;
   global $trans_hosting;
   global $modified_source;
   global $news_flash;
   global $news_counter;
   global $script_lang;
   global $trans_old_version;

# Close the body table

  echo "  </td> <!--- end of page contents --->\n";

# Insert news flash

  if(!strcmp($script_file, "index.php"))
  {
?>

  <!--- news flash column --->
  <td></td>
  <td bgcolor="#e0e0ff" valign="top" width="20%">
    <table width="100%" cellpadding="10"><tr><td>
<?php
  echo "      <font size=\"-1\"><b>$trans_news</b></font>\n";
  echo "      <a href=\"http://dvdisaster.net/$script_lang/feed/atom.xml\"><img src=\"../images/atom16.png\" border=></a>\n";
?>
      <table width="100%" cellpadding="0" cellspacing="0">
         <tr bgcolor="#000000">
	   <td width="100%"><img width=1 height=1 alt=""></td>
	 </tr>
      </table><p>
      <table width="100%">
        <tr><td>
<?php
	  $news_flash=1;
	  require("news.php");
?>
        </td></tr>
      </table>
    </table>
  </td> <!--- end of news flash column--->
<?php
  }
# Reference to our hoster
/*
 </tr>
 <tr valign="bottom">
   <td bgcolor="#f0f0f0" align="center">
<?php echo "     <font size=\"-2\">$trans_hosting</font><br>\n"; ?>
      <img src="../images/mokelbude.png"
           width="125" height="37" border="0" alt="insider">
   </td>
   <td></td>
   <td></td>
*/
?>

 </tr>
 <tr valign="bottom">
   <td bgcolor="#f0f0f0">
<?php 
   $old_lang = strcmp($script_lang, "ru") ? $script_lang : "en";
   echo "      <table cellpadding=\"10\"><tr><td><a href=\"http://dvdisaster.net/legacy/$old_lang/index.html\"><font size=\"-1\">$trans_old_version</font></a></td></tr></table>\n"; 
?>
   </td>
   <td></td>
   <td></td>

 </tr>
</table> <!--- end of main body table --->

<?php
# Create the footer
?>

<!--- Page footer --->
<table width="100%" cellpadding="0" border="0">
 <tr><td colspan="2" width="100%" height="10"><img width=1 height=1 alt=""></td>
 </tr>
 <tr bgcolor="#000000"><td colspan="2" width="100%"><img width=1 height=1 alt=""></td>
 </tr>
 <tr>
  <td align="center">
   <font size="-1">
<?php
   echo "     <i> $trans_copyright<br>\n";
   echo "         $trans_fdl\n";
?>
    </i>
   </font>
  </td>
 </tr>
 <tr bgcolor="#000000"><td colspan="2" width="100%"><img width=1 height=1 alt=""></td>
 </tr>
</table>
</body>
</html>
<?php
}

#
# Special headings etc
#

function howto_headline($headline, $subtitle, $image)
{
   echo "<table width=\"100%\" bgcolor=\"#f0f0f0\">\n";
   echo "<tr><td><font size=\"+1\"><b>$headline</b></font></td>\n";
   echo "<td rowspan=\"2\" align=\"right\"><img src=\"$image\"></td></tr>\n";
   echo "<tr><td><i>$subtitle</i></td></tr>\n";
   echo "</table><p>\n";
}
?>
